ragstack/colbert/astra_retriever.py

```diff
@@ -0,0 +1,162 @@
+    return max_sim
+
+
+class ColbertAstraRetriever:
+    astra: AstraDB
+    colbertEmbeddings: ColbertTokenEmbeddings
+    verbose: bool
+    is_cuda: bool = False
+
+    class Config:
+        arbitrary_types_allowed = True
+
+    def __init__(
+        self,
+        astraDB: AstraDB,
+        colbertEmbeddings: ColbertTokenEmbeddings,
+        verbose: bool = False,
+        **kwargs,
+    ):
+        # initialize pydantic base model
+        self.astra = astraDB
+        self.colbertEmbeddings = colbertEmbeddings
+        self.verbose = verbose
+        self.is_cuda = torch.cuda.is_available()
+
+    def retrieve(self, query: str, k: int = 10, query_maxlen: int = 64, **kwargs):
+        #
+        # if the query has fewer than a predefined number of of tokens Nq,
+        # colbertEmbeddings will pad it with BERT special [mast] token up to length Nq.
+        #
+        query_encodings = self.colbertEmbeddings.encode_query(
+            query, query_maxlen=query_maxlen
+        )
+
+        # the min of query_maxlen is 32
+        top_k = max(math.floor(len(query_encodings) / 2), 16)
+        if self.verbose:
+            logging.info(f"query length {len(query)} embeddings top_k: {top_k}")
+
+        # find the most relevant documents
+        docparts = set()
+        doc_futures = []
+        for qv in query_encodings:
+            # per token based retrieval
+            doc_future = self.astra.session.execute_async(
+                self.astra.query_colbert_ann_stmt, [list(qv), top_k]
+            )
+            doc_futures.append(doc_future)
+
+        for future in doc_futures:
+            rows = future.result()
+            docparts.update((row.title, row.part) for row in rows)
+
+        # score each document
+        scores = {}
+        futures = []
+        for title, part in docparts:
+            future = self.astra.session.execute_async(
+                self.astra.query_colbert_parts_stmt, [title, part]
+            )
+            futures.append((future, title, part))
+
+        for title, part in docparts:
+            # blocking call until the future is done
+            rows = future.result()
+            # find all the found parts so that we can do max similarity search
+            embeddings_for_part = [tensor(row.bert_embedding) for row in rows]
+            # score based on The function returns the highest similarity score
+            # (i.e., the maximum dot product value) between the query vector and any of the embedding vectors in the list.
+            scores[(title, part)] = sum(
+                max_similarity_torch(qv, embeddings_for_part, self.is_cuda)
+                for qv in query_encodings
+            )
+        # load the source chunk for the top k documents
+        docs_by_score = sorted(scores, key=scores.get, reverse=True)[:k]
+        answers = []
+        rank = 1
+        for title, part in docs_by_score:
+            rs = self.astra.session.execute(
```

Comment:
here there's room for improvement by doing it async 

then sort answers again before returning 

no need to do it now

Reply:
adde async execution